### PR TITLE
Simplify browser integration including cross-platform proxy code

### LINKED ISSUE #4121: Browser integration doesn't respect XDG_CONFIG_HOME
I migrated from an Ubuntu 16.04 machine to a new machine with 18.04. KeepassXC is working, and the browser integration for Firefox as well, but for Chrome it is not. It was working before migration.

I've enabled the browser support in the KeepassXC settings and installed the Chrome Extension. Both should be up to date.

If I click on the KeepassXC icon in Chrome it shows:

```
KeePassXC-Browser has encountered an error:

Cannot connect to KeePassXC. Check that browser integration is enabled in KeePassXC settings.
```

If I click  "Reload", the error message changes to
```
Key exchange was not successful.
```

In the extension settings, under "connected databases", there are no entries, and if I click "Connect", nothing happens. After migration, there were two entries, but since it wasn't working, I have removed them in the meantime.

One thing that is different to the previous machine, is that now I have `XDG_CONFIG_HOME` set up, and I copied my chrome user profile to the new location. It is now in `~/bionic/.config/google-chrome/Profile 1`, whereas before it was in the more default `~/.config/google-chrome/Profile 1`. Could that be a reason for the issue? Note that Firefox doesn't respect `XDG_CONFIG_HOME` and the firefox profile is still under `~/.mozilla` like on the old machine.

## Debug info
There seem to be no keepassxc related error messages in the JS console (depending on the site I have open no messages at all)
KeePassXC - 2.5.1
KeePassXC-Browser - 1.5.4
Operating system: Linux
Browser: Chrome

### PR DESCRIPTION
**Pull request by droidmonkey**
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )
* Only allow use of the proxy application when communicating with browser extension

* Complete refactor of Browser Integration classes
  * Moved browser service entry point to BrowserService class instead of NativeMessagingHost. BrowserService now coordinates the communication to/from clients.
  * Moved settings page definition out of MainWindow
  * Decoupled BrowserService from DatabaseTabWidget
  * Reduced complexity of various functions and cleaned the ABI (public vs private).
  * Eliminated BrowserClients class, moved functionality into the BrowserService
  * Renamed HostInstaller to NativeMessageInstaller and renamed NativeMessageHost to BrowserHost.

Fix #4121

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Updated the browser tests. The main purpose of this refactor is to eliminate platform-specific code for browser integration in the proxy and KeePassXC. The secondary purpose of the refactor is to attempt to address crashes and freezes that happen when integration is enabled. The most likely culprit was mishandling of mutex's.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)
- ✅ Breaking change (causes existing functionality to change)
- ✅ Refactor (significant modification to existing code)

### CODE REVIEW

File: src/gui/MainWindow.cpp [hunk 1]
```diff
@@ -185,13 +132,19 @@ MainWindow::MainWindow()
...
+            &DatabaseTabWidget::activateDatabaseChanged,
+            browserService(),
+            &BrowserService::activeDatabaseChanged);
 #endif
 
 #ifdef WITH_XC_SSHAGENT
     connect(sshAgent(), SIGNAL(error(QString)), this, SLOT(showErrorMessage(QString)));
     connect(sshAgent(), SIGNAL(enabledChanged(bool)), this, SLOT(agentEnabled(bool)));
-    m_ui->settingsWidget->addSettingsPage(new AgentSettingsPage(m_ui->tabWidget));
+    m_ui->settingsWidget->addSettingsPage(new AgentSettingsPage());
```
Comment (droidmonkey) [inline src/gui/MainWindow.cpp:147]: The tab widget isn't even used by SSH Agent!

File: src/browser/BrowserService.cpp [hunk 2]
```diff
@@ -58,34 +60,45 @@ const QString BrowserService::OPTION_ONLY_HTTP_AUTH = QStringLiteral("BrowserOnl
...
+    connect(m_browserHost, &BrowserHost::clientMessageReceived, this, &BrowserService::processClientMessage);
+    setEnabled(browserSettings()->isEnabled());
+}
+
+BrowserService* BrowserService::instance()
+{
+    return s_browserService;
+}
+
+void BrowserService::setEnabled(bool state)
```
Comment (varjolintu) [inline src/browser/BrowserService.cpp:82]: Maybe just `enabled` instead of `state` because it's a boolean, not a possible multi-value state.

**Comment by varjolintu** [comment]
We could also make the support for `XDG_CONFIG_HOME`.
https://github.com/keepassxreboot/keepassxc/issues/4121

File: src/browser/BrowserShared.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#include "config-keepassx.h"
+
+#include <QProcessEnvironment>
+#include <QStandardPaths>
+
+namespace Browser
+{
+    QString localServerPath()
+    {
+        const QString serverPath = QStringLiteral("/kpxc_browserserver");
```
[the remark this answers is not in the document]
Comment (phoerious) [reply src/browser/BrowserShared.cpp:28]: If we're already changing this, what about changing it to `org.keepassxc.KeePassXC.BrowserServer`?

File: src/browser/BrowserShared.cpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#if defined(KEEPASSXC_DIST_SNAP)
+        return QProcessEnvironment::systemEnvironment().value("SNAP_USER_COMMON") + serverPath;
+#elif defined(Q_OS_UNIX) && !defined(Q_OS_MACOS)
+        // Use XDG_RUNTIME_DIR instead of /tmp if it's available
+        QString path = QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation);
+        return path.isEmpty() ? QStandardPaths::writableLocation(QStandardPaths::TempLocation) + serverPath
+                              : path + serverPath;
+#elif defined(Q_OS_WIN)
+        return serverPath;
```
Comment (phoerious) [inline src/browser/BrowserShared.cpp:37]: Should add a comment that this is the name a named pipe, not a socket path.

File: src/browser/NativeMessageInstaller.cpp [hunk 5]
```diff
@@ -234,55 +212,27 @@ QString HostInstaller::getPath(SupportedBrowsers browser) const
...
-            path.append(".exe");
-#endif
-        }
-    } else {
-        path = QFileInfo(QCoreApplication::applicationFilePath()).absoluteFilePath();
-    }
+    path = QCoreApplication::applicationDirPath() + QStringLiteral("/keepassxc-proxy");
 #ifdef Q_OS_WIN
+    path.append(QStringLiteral(".exe"));
     path.replace("/", "\\");
```
Comment (phoerious) [inline src/browser/NativeMessageInstaller.cpp:234]: `QDir::toNativeSeparators()`

File: src/proxy/CMakeLists.txt [hunk 6]
```diff
@@ -15,16 +14,14 @@
 #  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 if(WITH_XC_BROWSER)
-    include_directories(${BROWSER_SOURCE_DIR})
-
     set(proxy_SOURCES
-        ../core/Alloc.cpp
```
Comment (phoerious) [inline src/proxy/CMakeLists.txt:21]: I suppose you removed this because we're linking statically?

File: src/proxy/NativeMessagingProxy.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+#ifdef Q_OS_WIN
+    setmode(fileno(stdin), _O_BINARY);
+    setmode(fileno(stdout), _O_BINARY);
+#endif
+
+    QtConcurrent::run([this] {
+        while (std::cin.good()) {
+            if (std::cin.peek() != EOF) {
+                unsigned int length = 0;
+                for (int i = 0; i < 4; ++i) {
```
Comment (phoerious) [inline src/proxy/NativeMessagingProxy.cpp:55]: I would make it explicit with `sizeof(unsigned int)`.

File: src/proxy/NativeMessagingProxy.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                    length |= getchar() << (i * 8);
+                }
+
+                QString msg;
+                for (unsigned int i = 0; i < length; ++i) {
+                    msg += getchar();
+                }
```
Comment (phoerious) [inline src/proxy/NativeMessagingProxy.cpp:62]: `QStringBuilder`.

File: src/proxy/NativeMessagingProxy.cpp [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+{
+    if (m_localSocket && m_localSocket->state() == QLocalSocket::ConnectedState) {
+        m_localSocket->write(msg.toUtf8(), msg.length());
+        m_localSocket->flush();
+    }
+}
+
+void NativeMessagingProxy::setupLocalSocket()
+{
+    m_localSocket = new QLocalSocket();
```
Comment (phoerious) [inline src/proxy/NativeMessagingProxy.cpp:84]: Any reason why this can't be a `QScopedPointer`?

File: src/proxy/NativeMessagingProxy.cpp [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+void NativeMessagingProxy::newLocalMessage()
+{
+    auto msg = m_localSocket->readAll();
+    if (!msg.isEmpty()) {
+        uint len = msg.size();
+        std::cout << char(((len >> 0) & 0xFF)) << char(((len >> 8) & 0xFF)) << char(((len >> 16) & 0xFF))
+                  << char(((len >> 24) & 0xFF));
+        std::cout << msg.toStdString() << std::flush;
```
Comment (phoerious) [inline src/proxy/NativeMessagingProxy.cpp:99]: Too many magic numbers. No idea what's going on here. And why are we using `iostream` instead of `QTextStream`?

Comment (varjolintu) [reply src/proxy/NativeMessagingProxy.cpp:99]: The message length must be passed at the beginning as a 32-bit number (native byte order). If we can replace this with a QTextStream, that would be nice.

File: src/proxy/CMakeLists.txt (hunk 6 above)
Comment (droidmonkey) [reply src/proxy/CMakeLists.txt:21]: No I should not have removed this file, adding back in. Good catch.

File: src/proxy/NativeMessagingProxy.cpp (hunk 8 above)
Comment (droidmonkey) [reply src/proxy/NativeMessagingProxy.cpp:62]: That is an internal class and doesn't really make sense in this context. Instead I went with QString::reserve(length) and QString::append instead of the `+` operator.

File: src/proxy/NativeMessagingProxy.cpp (hunk 9 above)
Comment (droidmonkey) [reply src/proxy/NativeMessagingProxy.cpp:84]: Sure can!

File: src/proxy/NativeMessagingProxy.cpp (hunk 10 above)
Comment (droidmonkey) [reply src/proxy/NativeMessagingProxy.cpp:99]: This is the only way I can get it to work, this is from the original implementation prior to refactor.

File: src/browser/BrowserShared.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/browser/BrowserShared.cpp]: Good idea, all our pipes should follow this convention including the single instance.

File: src/proxy/NativeMessagingProxy.cpp (hunk 10 above)
Comment (phoerious) [reply src/proxy/NativeMessagingProxy.cpp:100]: QTextStream should work just like std::iostream.

Comment (droidmonkey) [reply src/proxy/NativeMessagingProxy.cpp:100]: This works great on all platforms and I don't really want to mess with it anymore.

Comment (phoerious) [reply src/proxy/NativeMessagingProxy.cpp:100]: I just don't like mixing libraries too much and we're using QTextStream everywhere else. It should really be a drop-in replacement. And please give some names to the magic numbers or add a short comment above.

File: src/proxy/NativeMessagingProxy.cpp (hunk 8 above)
Comment (phoerious) [reply src/proxy/NativeMessagingProxy.cpp]: Also fine.

File: src/proxy/NativeMessagingProxy.cpp (hunk 10 above)
Comment (droidmonkey) [reply src/proxy/NativeMessagingProxy.cpp:100]: They aren't magic, just bit shifting 0, 1, 2, 3 bytes 😛

Comment (droidmonkey) [reply src/proxy/NativeMessagingProxy.cpp]: I made it sexy for yah

Comment (phoerious) [reply src/proxy/NativeMessagingProxy.cpp]: I can see that you're shifting bytes, but it's not immediately obvious why and what the expected outcome is.

File: src/browser/BrowserShared.cpp [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+#include "config-keepassx.h"
+
+#include <QProcessEnvironment>
+#include <QStandardPaths>
+
+namespace BrowserShared
+{
+    QString localServerPath()
+    {
+        const QString serverName = QStringLiteral("/org.keepassxc.KeePassXC.BrowserServer");
```
Comment (phoerious) [inline src/browser/BrowserShared.cpp:28]: If we merge my config refactor branch first, you could get the prefix dynamically from a global application property.

File: src/browser/NativeMessageInstaller.cpp [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+#include <QMessageBox>
+#include <QProcessEnvironment>
+#include <QSettings>
+#include <QStandardPaths>
+
+using namespace BrowserShared;
+
+namespace
+{
+    const QString HOST_NAME = QStringLiteral("org.keepassxc.keepassxc_browser");
```
Comment (phoerious) [inline src/browser/NativeMessageInstaller.cpp:38]: I suppose this couldn't be changed to `org.keepassxc.KeePassXC.KeePassXCBrowser` without breaking stuff, could it?

File: src/browser/NativeMessageInstaller.cpp [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+    case SupportedBrowsers::VIVALDI:
+        return TARGET_DIR_VIVALDI;
+    case SupportedBrowsers::TOR_BROWSER:
+        return TARGET_DIR_TOR_BROWSER;
+    case SupportedBrowsers::BRAVE:
+        return TARGET_DIR_BRAVE;
+    case SupportedBrowsers::EDGE:
+        return TARGET_DIR_EDGE;
+    default:
+        return QString();
```
Comment (phoerious) [inline src/browser/NativeMessageInstaller.cpp:163]: `return {}`

File: src/browser/NativeMessageInstaller.cpp [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+    case SupportedBrowsers::VIVALDI:
+        return "vivaldi";
+    case SupportedBrowsers::TOR_BROWSER:
+        return "tor-browser";
+    case SupportedBrowsers::BRAVE:
+        return "brave";
+    case SupportedBrowsers::EDGE:
+        return "edge";
+    default:
+        return QString();
```
Comment (phoerious) [inline src/browser/NativeMessageInstaller.cpp:192]: `return {}`

File: src/browser/NativeMessageInstaller.cpp [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+    case SupportedBrowsers::FIREFOX:
+        return "firefox";
+    case SupportedBrowsers::VIVALDI:
+        return "vivaldi";
+    case SupportedBrowsers::TOR_BROWSER:
+        return "tor-browser";
+    case SupportedBrowsers::BRAVE:
+        return "brave";
+    case SupportedBrowsers::EDGE:
+        return "edge";
```
Comment (phoerious) [inline src/browser/NativeMessageInstaller.cpp:190]: `QStringLiteral`

File: src/browser/NativeMessageInstaller.cpp [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+{
+    QString basePath;
+#if defined(Q_OS_WIN)
+    // If portable settings file exists save the JSON scripts to the application folder
+    if (QFile::exists(QCoreApplication::applicationDirPath() + QStringLiteral("/keepassxc.ini"))) {
+        basePath = QCoreApplication::applicationDirPath();
+    } else {
+        basePath = QStandardPaths::writableLocation(QStandardPaths::DataLocation);
+    }
+    return QString("%1/%2_%3.json").arg(basePath, HOST_NAME, getBrowserName(browser));
```
Comment (phoerious) [inline src/browser/NativeMessageInstaller.cpp:212]: `QStringLiteral`

File: src/browser/NativeMessageInstaller.cpp [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+    } else if (browser == SupportedBrowsers::FIREFOX) {
+        // Firefox stores its config in ~/
+        basePath = QDir::homePath();
+    } else {
+        basePath = QStandardPaths::writableLocation(QStandardPaths::ConfigLocation);
+    }
+#else
+    basePath = QDir::homePath();
+#endif
+    return QString("%1%2/%3.json").arg(basePath, getTargetPath(browser), HOST_NAME);
```
Comment (phoerious) [inline src/browser/NativeMessageInstaller.cpp:226]: `QStringLiteral`

File: src/browser/NativeMessageInstaller.cpp [hunk 18]
```diff
@@ -0,0 +1,313 @@
...
+        return browserSettings()->customProxyLocation();
+    }
+
+    QString path;
+#ifdef KEEPASSXC_DIST_APPIMAGE
+    path = QProcessEnvironment::systemEnvironment().value("APPIMAGE");
+#else
+    path = QCoreApplication::applicationDirPath() + QStringLiteral("/keepassxc-proxy");
+#ifdef Q_OS_WIN
+    path.append(QStringLiteral(".exe"));
```
Comment (phoerious) [inline src/browser/NativeMessageInstaller.cpp:247]: However, I don't think it's needed here. `QString::append()` has a `const char*` overload. 🤣

File: src/proxy/NativeMessagingProxy.cpp [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+        std::cout.write(reinterpret_cast<char*>(&len), sizeof(len));
+        
+        // Write the message and flush the stream
+        std::cout << msg.toStdString() << std::flush;
+    }
+}
+
+void NativeMessagingProxy::deleteSocket()
+{
+    m_localSocket->deleteLater();
```
Comment (phoerious) [inline src/proxy/NativeMessagingProxy.cpp:108]: I actually don't know what happens to the pointer itself if you do this on a `QScopedPointer`. If `deleteLater()` is desired, you should be using the `QScopedPointerDeleteLater` handler and simply call `reset()` here.

File: src/browser/NativeMessageInstaller.cpp (hunk 12 above)
Comment (varjolintu) [reply src/browser/NativeMessageInstaller.cpp:38]: We cannot change this. From the specification: "This name can only contain lowercase alphanumeric characters, underscores and dots. The name cannot start or end with a dot, and a dot cannot be followed by another dot."

File: src/browser/BrowserShared.cpp (hunk 11 above)
Comment (droidmonkey) [reply src/browser/BrowserShared.cpp:28]: Perfect

File: src/browser/NativeMessageInstaller.cpp (hunk 18 above)
Comment (droidmonkey) [reply src/browser/NativeMessageInstaller.cpp:247]: Consistency ftw!

File: src/proxy/NativeMessagingProxy.cpp (hunk 19 above)
Comment (droidmonkey) [reply src/proxy/NativeMessagingProxy.cpp:108]: Since we moved to scoped pointer this line is indeed redundant. The proxy is designed to exit immediately upon disconnect from either keepassxc or browser side. The former should not happen but @varjolintu is working on that with his reconnect work.

File: src/browser/NativeMessageInstaller.cpp (hunk 12 above)
Comment (phoerious) [reply src/browser/NativeMessageInstaller.cpp:38]: Should be org.keepassxc.keepassxc.browser then, unless it breaks something, in that case just leave it as is.

Comment (varjolintu) [reply src/browser/NativeMessageInstaller.cpp:38]: Leave it as it is.

File: src/browser/BrowserShared.cpp (hunk 11 above)
Comment (phoerious) [reply src/browser/BrowserShared.cpp:28]: What's the status of this?

Comment (droidmonkey) [reply src/browser/BrowserShared.cpp:28]: I am actually not sure what you were going after with this comment. Can you rephrase or implement yourself? I pushed my final changes that addresses everything else.

Comment (phoerious) [reply src/browser/BrowserShared.cpp:28]: You wanted to get this name dynamically from the app property that is set globally in main.cpp.
